{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "## Setup notebook"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 257,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-06-05T01:29:31.405416Z",
     "start_time": "2018-06-05T01:29:31.392792Z"
    }
   },
   "outputs": [
    {
     "data": {
      "application/javascript": [
       "IPython.notebook.set_autosave_interval(10000)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Autosaving every 10 seconds\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import math\n",
    "#from scipy import stats\n",
    "\n",
    "# plotting\n",
    "#import matplotlib as mpl\n",
    "\n",
    "#import pickle\n",
    "\n",
    "import re\n",
    "#from collections import Counter\n",
    "from collections import defaultdict\n",
    "\n",
    "import yaml\n",
    "import builtins\n",
    "\n",
    "# jupyter notebook amenities\n",
    "#%matplotlib inline\n",
    "%autosave 10"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Read in data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Drug and Biomarkers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 270,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-06-05T01:40:37.274593Z",
     "start_time": "2018-06-05T01:40:37.250712Z"
    },
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "oncokb_yaml_file = '../OncoKB-Drug-Information/oncokb.yaml'\n",
    "oncokb_text_file = '../OncoKB-Drug-Information/allActionableVariants.txt'\n",
    "\n",
    "#with open(oncokb_yaml_file, 'r') as f:\n",
    "#    oncokb = yaml.load(f)\n",
    "\n",
    "oncokb_table = pd.read_csv(oncokb_text_file, sep='\\t', header=0, index_col=None)\n",
    "\n",
    "\n",
    "#oncokb = defaultdict(list)\n",
    "#for i, row in oncokb_table.iterrows():\n",
    "#    # Save desired values for a given alteration\n",
    "#    v = {'Gene': row.Gene,\n",
    "#         'Alteration': row.Alteration,\n",
    "#         'Subtype': row['Cancer Type'],\n",
    "#         'Level': row.Level}\n",
    "#    # Save for each drug listed for an alteration\n",
    "#    drugs = row['Drugs(s)'].split(', ')\n",
    "#    for drug in drugs:\n",
    "#        oncokb[drug].append(v)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Mutation data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-05-29T21:31:24.756279Z",
     "start_time": "2018-05-29T21:05:49.077576Z"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/vfriedl/anaconda/lib/python3.6/site-packages/IPython/core/interactiveshell.py:2717: DtypeWarning: Columns (4,59) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  interactivity=interactivity, compiler=compiler, result=result)\n"
     ]
    }
   ],
   "source": [
    "tcga_mutation_filename = '../../../../TCGA_Pancanatlas_SNVs_out.tsv'\n",
    "tcga_mutation_data = pd.read_csv(tcga_mutation_filename, sep='\\t', header=0, index_col=None)\n",
    "\n",
    "print(tcga_mutation_data.shape)\n",
    "print(list(tcga_mutation_data))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### CNV data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-05-29T21:48:50.055542Z",
     "start_time": "2018-05-29T21:47:17.878604Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(25128, 10715)\n"
     ]
    }
   ],
   "source": [
    "tcga_CNV_filename = '../../../../TCGA_Pancanatlas_CNV_Thresholded_out.tsv'\n",
    "tcga_cnv_data = pd.read_csv(tcga_CNV_filename, sep='\\t', header=0, index_col=0)\n",
    "\n",
    "print(tcga_cnv_data.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Fusion data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### tumorfusions.org"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 248,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-06-05T01:21:04.246088Z",
     "start_time": "2018-06-05T01:21:04.028977Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(17754, 27)\n",
      "5517\n",
      "5499\n"
     ]
    }
   ],
   "source": [
    "tumorfusions_filename = \"../../../../fusion_data/pancanfus_newline.txt\"\n",
    "tcga_fusion_data = pd.read_csv(tumorfusions_filename, sep='\\t', header=0)\n",
    "\n",
    "tcga_fusion_data[\"sampleId\"] = [\"TCGA-\"+x[0:2]+\"-\"+x[3:7]+\"-\"+x[8:] for x in tcga_fusion_data[\"sampleId\"]]\n",
    "\n",
    "print(tcga_fusion_data.shape)\n",
    "print(len(set(tcga_fusion_data[\"sampleId\"])))\n",
    "samples_in_fusion_data = [x[:15] for x in tcga_fusion_data[\"sampleId\"]]\n",
    "print(len(set(samples_in_fusion_data)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Clinical data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-05-29T21:49:04.884857Z",
     "start_time": "2018-05-29T21:49:03.176538Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['bcr_patient_barcode', 'acronym', 'gender', 'vital_status', 'days_to_birth', 'days_to_death', 'days_to_last_followup', 'days_to_initial_pathologic_diagnosis', 'age_at_initial_pathologic_diagnosis', 'icd_10', 'tissue_retrospective_collection_indicator', 'icd_o_3_histology', 'tissue_prospective_collection_indicator', 'history_of_neoadjuvant_treatment', 'icd_o_3_site', 'tumor_tissue_site', 'new_tumor_event_after_initial_treatment', 'radiation_therapy', 'race', 'project_code', 'prior_dx', 'disease_code', 'ethnicity', 'informed_consent_verified', 'person_neoplasm_cancer_status', 'patient_id', 'year_of_initial_pathologic_diagnosis', 'histological_type', 'tissue_source_site', 'form_completion_date', 'pathologic_T', 'pathologic_M', 'clinical_M', 'pathologic_N', 'system_version', 'pathologic_stage', 'stage_other', 'clinical_stage', 'clinical_T', 'clinical_N', 'extranodal_involvement', 'postoperative_rx_tx', 'primary_therapy_outcome_success', 'lymph_node_examined_count', 'primary_lymph_node_presentation_assessment', 'initial_pathologic_diagnosis_method', 'number_of_lymphnodes_positive_by_he', 'eastern_cancer_oncology_group', 'anatomic_neoplasm_subdivision', 'residual_tumor', 'histological_type_other', 'init_pathology_dx_method_other', 'karnofsky_performance_score', 'neoplasm_histologic_grade', 'height', 'weight', 'history_of_radiation_metastatic_site', 'days_to_patient_progression_free', 'days_to_first_response', 'days_to_first_partial_response', 'days_to_first_complete_response', 'days_to_tumor_progression', 'er_estimated_duration_response', 'er_disease_extent_prior_er_treatment', 'er_solid_tumor_response_documented_type', 'er_solid_tumor_response_documented_type_other', 'er_response_type', 'history_of_radiation_primary_site', 'history_prior_surgery_type', 'patient_progression_status', 'history_prior_surgery_type_other', 'history_prior_surgery_indicator', 'field', 'molecular_abnormality_results_other', 'molecular_abnormality_results', 'number_of_lymphnodes_positive_by_ihc', 'tobacco_smoking_history', 'number_pack_years_smoked', 'stopped_smoking_year', 'performance_status_scale_timing', 'laterality', 'targeted_molecular_therapy', 'year_of_tobacco_smoking_onset', 'anatomic_neoplasm_subdivision_other', 'patient_death_reason', 'tumor_tissue_site_other', 'menopause_status', 'age_began_smoking_in_years', 'margin_status', 'kras_gene_analysis_performed', 'kras_mutation_found', 'death_cause_text', 'lactate_dehydrogenase_result', 'days_to_sample_procurement', 'hbv_test', 'on_haart_therapy_at_cancer_diagnosis', 'hcv_test', 'prior_aids_conditions', 'on_haart_therapy_prior_to_cancer_diagnosis', 'kshv_hhv8_test', 'days_to_hiv_diagnosis', 'hiv_status', 'hiv_rna_load_at_diagnosis', 'cdc_hiv_risk_group', 'history_of_other_malignancy', 'history_immunosuppresive_dx', 'nadir_cd4_counts', 'history_relevant_infectious_dx_other', 'history_relevant_infectious_dx', 'cd4_counts_at_diagnosis', 'history_immunological_disease_other', 'hpv_test', 'history_immunosuppressive_dx_other', 'history_immunological_disease', 'lost_follow_up', 'venous_invasion', 'lymphatic_invasion', 'perineural_invasion_present', 'her2_erbb_method_calculation_method_text', 'her2_immunohistochemistry_level_result', 'breast_carcinoma_immunohistochemistry_pos_cell_score', 'distant_metastasis_present_ind2', 'her2_erbb_pos_finding_fluorescence_in_situ_hybridization_calculation_method_text', 'breast_carcinoma_immunohistochemistry_progesterone_receptor_pos_finding_scale', 'breast_carcinoma_primary_surgical_procedure_name', 'er_level_cell_percentage_category', 'breast_carcinoma_progesterone_receptor_status', 'breast_carcinoma_surgical_procedure_name', 'breast_neoplasm_other_surgical_procedure_descriptive_text', 'er_detection_method_text', 'pos_finding_progesterone_receptor_other_measurement_scale_text', 'her2_erbb_pos_finding_cell_percent_category', 'her2_and_centromere_17_positive_finding_other_measurement_scale_text', 'pos_finding_metastatic_breast_carcinoma_estrogen_receptor_other_measuremenet_scale_text', 'metastatic_breast_carcinoma_progesterone_receptor_level_cell_percent_category', 'metastatic_breast_carcinoma_pos_finding_progesterone_receptor_other_measure_scale_text', 'fluorescence_in_situ_hybridization_diagnostic_procedure_chromosome_17_signal_result_range', 'first_recurrent_non_nodal_metastatic_anatomic_site_descriptive_text', 'first_nonlymph_node_metastasis_anatomic_site', 'immunohistochemistry_positive_cell_score', 'axillary_lymph_node_stage_method_type', 'axillary_lymph_node_stage_other_method_descriptive_text', 'metastatic_breast_carcinoma_progesterone_receptor_status', 'metastatic_breast_carcinoma_immunohistochemistry_pr_pos_cell_score', 'metastatic_breast_carcinoma_immunohistochemistry_er_pos_cell_score', 'metastatic_breast_carcinoma_lab_proc_her2_neu_immunohistochemistry_receptor_status', 'metastatic_breast_carcinoma_her2_neu_chromosone_17_signal_ratio_value', 'surgical_procedure_purpose_other_text', 'metastatic_breast_carcinoma_her2_erbb_pos_finding_fluorescence_in_situ_hybridization_calculation_method_text', 'metastatic_breast_carcinoma_her2_erbb_pos_finding_cell_percent_category', 'breast_cancer_surgery_margin_status', 'metastatic_breast_carcinoma_her2_erbb_method_calculation_method_text', 'breast_carcinoma_estrogen_receptor_status', 'her2_neu_and_centromere_17_copy_number_analysis_input_total_number_count', 'pgr_detection_method_text', 'pos_finding_her2_erbb2_other_measurement_scale_text', 'breast_carcinoma_immunohistochemistry_er_pos_finding_scale', 'positive_finding_estrogen_receptor_other_measurement_scale_text', 'metastatic_breast_carcinoma_erbb2_immunohistochemistry_level_result', 'metastatic_breast_carcinoma_pos_finding_other_scale_measurement_text', 'her2_neu_chromosone_17_signal_ratio_value', 'cytokeratin_immunohistochemistry_staining_method_micrometastasis_indicator', 'metastatic_breast_carcinoma_estrogen_receptor_detection_method_text', 'metastatic_breast_carcinoma_lab_proc_her2_neu_in_situ_hybridization_outcome_type', 'metastatic_breast_carcinoma_pos_finding_her2_erbb2_other_measure_scale_text', 'lab_proc_her2_neu_immunohistochemistry_receptor_status', 'progesterone_receptor_level_cell_percent_category', 'metastatic_breast_carcinoma_estrogen_receptor_level_cell_percent_category', 'metastatic_breast_carcinoma_estrogen_receptor_status', 'lab_procedure_her2_neu_in_situ_hybrid_outcome_type', 'her2_neu_breast_carcinoma_copy_analysis_input_total_number', 'metastatic_breast_carcinoma_fluorescence_in_situ_hybridization_diagnostic_proc_centromere_17_signal_result_range', 'her2_neu_and_centromere_17_copy_number_metastatic_breast_carcinoma_analysis_input_total_number_count', 'her2_neu_metastatic_breast_carcinoma_copy_analysis_input_total_number', 'metastatic_breast_carcinoma_progesterone_receptor_detection_method_text', 'additional_pharmaceutical_therapy', 'additional_radiation_therapy', 'lymphovascular_invasion_present', 'pos_lymph_node_location_other', 'pos_lymph_node_location', 'location_in_lung_parenchyma', 'pulmonary_function_test_performed', 'pre_bronchodilator_fev1_fvc_percent', 'kras_mutation_result', 'post_bronchodilator_fev1_percent', 'pre_bronchodilator_fev1_percent', 'dlco_predictive_percent', 'egfr_mutation_performed', 'diagnosis', 'eml4_alk_translocation_result', 'egfr_mutation_result', 'post_bronchodilator_fev1_fvc_percent', 'eml4_alk_translocation_performed', 'eml4_alk_translocation_method', 'days_to_new_tumor_event_after_initial_treatment', 'hemoglobin_result', 'serum_calcium_result', 'platelet_qualitative_result', 'number_of_lymphnodes_positive', 'erythrocyte_sedimentation_rate_result', 'white_cell_count_result', 'malignant_neoplasm_metastatic_involvement_site', 'other_metastatic_involvement_anatomic_site', 'birth_control_pill_history_usage_category', 'alcohol_history_documented', 'frequency_of_alcohol_consumption', 'source_of_patient_death_reason', 'days_to_diagnostic_mri_performed', 'days_to_diagnostic_computed_tomography_performed', 'amount_of_alcohol_consumption_per_day', 'family_history_of_cancer', 'loss_expression_of_mismatch_repair_proteins_by_ihc_result', 'preoperative_pretreatment_cea_level', 'microsatellite_instability', 'number_of_loci_tested', 'history_of_colon_polyps', 'number_of_first_degree_relatives_with_cancer_diagnosis', 'number_of_abnormal_loci', 'circumferential_resection_margin', 'non_nodal_tumor_deposits', 'braf_gene_analysis_result', 'colon_polyps_present', 'kras_mutation_codon', 'braf_gene_analysis_performed', 'synchronous_colon_cancer_present', 'loss_expression_of_mismatch_repair_proteins_by_ihc', 'barretts_esophagus', 'city_of_procurement', 'reflux_history', 'country_of_procurement', 'antireflux_treatment_type', 'h_pylori_infection', 'hypertension', 'horm_ther', 'pln_pos_ihc', 'surgical_approach', 'pregnancies', 'diabetes', 'peritoneal_wash', 'pct_tumor_invasion', 'prior_tamoxifen_administered_usage_category', 'colorectal_cancer', 'total_aor-lnp', 'pln_pos_light_micro', 'total_pelv_lnp', 'total_pelv_lnr', 'aln_pos_light_micro', 'aln_pos_ihc', 'total_aor_lnr', 'post_surgical_procedure_assessment_thyroid_gland_carcinoma_status', 'metastatic_neoplasm_confirmed_diagnosis_method_text', 'metastatic_neoplasm_confirmed_diagnosis_method_name', 'prior_glioma', 'tumor_residual_disease', 'jewish_origin', 'distant_metastasis_anatomic_site', 'new_neoplasm_event_type', 'lymphnode_dissection_method_left', 'lymphnode_neck_dissection', 'egfr_amplication_status', 'hpv_status_by_ish_testing', 'presence_of_pathological_nodal_extracapsular_spread', 'p53_gene_analysis', 'hpv_status_by_p16_testing', 'lymphnode_dissection_method_right', 'eml4_alk_translocation_identified', 'egfr_mutation_identified', 'supratentorial_localization', 'mold_or_dust_allergy_history', 'motor_movement_changes', 'inherited_genetic_syndrome_found', 'seizure_history', 'asthma_history', 'first_diagnosis_age_asth_ecz_hay_fev_mold_dust', 'first_diagnosis_age_of_food_allergy', 'ldh1_mutation_found', 'tumor_location', 'ldh1_mutation_test_method', 'animal_insect_allergy_types', 'first_presenting_symptom', 'animal_insect_allergy_history', 'first_presenting_symptom_longest_duration', 'ldh1_mutation_tested', 'visual_changes', 'first_diagnosis_age_of_animal_insect_allergy', 'history_ionizing_rt_to_head', 'sensory_changes', 'food_allergy_history', 'mental_status_changes', 'hay_fever_history', 'preoperative_corticosteroids', 'preoperative_antiseizure_meds', 'family_history_of_primary_brain_tumor', 'inherited_genetic_syndrome_result', 'days_to_initial_score_performance_status_scale', 'food_allergy_types', 'headache_history', 'eczema_history', 'other_genotyping_outcome_lab_results_text', 'extrathyroid_carcinoma_present_extension_status', 'i_131_first_administered_dose', 'i_131_subsequent_administered_dose', 'i_131_total_administered_dose', 'primary_neoplasm_focus_type', 'i_131_total_administered_preparation_technique', 'person_lifetime_risk_radiation_exposure_indicator', 'ret_ptc_rearrangement_genotyping_outcome_lab_results_text', 'genotyping_results_gene_mutation_not_reported_reason', 'neoplasm_depth', 'genotype_analysis_performed_indicator', 'ras_family_gene_genotyping_outcome_lab_results_text', 'neoplasm_length', 'neoplasm_width', 'radiosensitizing_agent_administered_indicator', 'braf_gene_genotyping_outcome_lab_results_text', 'lymph_node_preoperative_scan_indicator', 'radiation_therapy_administered_preparation_technique_text', 'lymph_node_preoperative_assessment_diagnostic_imaging_type', 'patient_personal_medical_history_thyroid_gland_disorder_name', 'first_degree_relative_history_thyroid_gland_carcinoma_diagnosis_relationship_type', 'radiation_therapy_administered_dose_text', 'patient_personal_medical_history_thyroid_other_specify_text', 'primary_thyroid_gland_neoplasm_location_anatomic_site', 'lymphnodes_examined', 'zone_of_origin', 'days_to_bone_scan_performed', 'diagnostic_mri_performed', 'diagnostic_mri_result', 'psa_value', 'days_to_first_biochemical_recurrence', 'secondary_pattern', 'primary_pattern', 'diagnostic_ct_abd_pelvis_result', 'bone_scan_results', 'biochemical_recurrence', 'tumor_level', 'gleason_score', 'number_of_lymphnodes_examined', 'days_to_psa', 'tertiary_pattern', 'diagnostic_ct_abd_pelvis_performed', 'family_medical_history_relative_family_member_relationship_type', 'tumor_type', 'primary_tumor_multiple_present_ind', 'prior_systemic_therapy_type', 'melanoma_clark_level_value', 'melanoma_origin_skin_anatomic_site', 'malignant_neoplasm_mitotic_count_rate', 'breslow_depth_value', 'tumor_tissue_site.1', 'melanoma_ulceration_indicator', 'primary_neoplasm_melanoma_dx', 'new_tumor_dx_prior_submitted_specimen_dx', 'primary_anatomic_site_count', 'interferon_90_day_prior_excision_admin_indicator', 'primary_melanoma_at_diagnosis_count', 'days_to_submitted_specimen_dx', 'antireflux_treatment', 'number_of_relatives_with_stomach_cancer', 'family_history_of_stomach_cancer', 'state_province_country_of_procurement', 'specimen_collection_method_name', 'child_pugh_classification_grade', 'hist_of_non_mibc', 'occupation_primary_job', 'person_concomitant_prostate_carcinoma_occurrence_indicator', 'person_concomitant_prostate_carcinoma_pathologic_t_stage', 'cancer_diagnosis_cancer_type_icd9_text_name', 'non_mibc_tx', 'bladder_carcinoma_extracapsular_extension_status', 'chemical_exposure_text', 'resp_maint_from_bcg_admin_month_dur', 'maint_therapy_course_complete', 'induction_course_complete', 'person_occupation_years_number', 'person_primary_industry_text', 'disease_extracapsular_extension_ind-3', 'complete_response_observed', 'diagnosis_subtype', 'diagnosis_age', 'mibc_90day_post_resection_bcg', 'person_occupation_description_text', 'surgical_procedure_name_other_specify_text', 'relative_family_cancer_history_ind_3', 'vascular_tumor_cell_invasion_type', 'viral_hepatitis_serology', 'liver_fibrosis_ishak_score_category', 'ablation_embolization_tx_adjuvant', 'laboratory_procedure_prothrombin_time_result_value', 'laboratory_procedure_creatinine_result_lower_limit_of_normal_value', 'laboratory_procedure_alpha_fetoprotein_outcome_value', 'laboratory_procedure_alpha_fetoprotein_outcome_upper_limit_of_normal_value', 'laboratory_procedure_alpha_fetoprotein_outcome_lower_limit_of_normal_value', 'laboratory_procedure_albumin_result_upper_limit_of_normal_value', 'laboratory_procedure_albumin_result_specified_value', 'laboratory_procedure_creatinine_result_upper_limit_of_normal_value', 'laboratory_prcoedure_platelet_result_upper_limit_of_normal_value', 'lab_procedure_platelet_result_specified_value', 'days_to_definitive_surgical_procedure_performed', 'hematology_serum_creatinine_laboratory_result_value_in_mg_dl', 'history_hepato_carcinoma_risk_factor', 'history_hepato_carcinoma_risk_factor_other', 'cancer_diagnosis_first_degree_relative_number', 'laboratory_prcoedure_platelet_result_lower_limit_of_normal_value', 'laboratory_procedure_international_normalization_ratio_result_lower_limit_of_normal_value', 'laboratory_procedure_albumin_result_lower_limit_of_normal_value', 'laboratory_procedure_total_bilirubin_result_specified_lower_limit_of_normal_value', 'laboratory_procedure_total_bilirubin_result_specified_upper_limit_of_normal_value', 'laboratory_procedure_total_bilirubin_result_upper_limit_normal_value', 'adjacent_hepatic_tissue_inflammation_extent_type', 'laboratory_procedure_international_normalization_ratio_result_upper_limit_of_normal_value', 'mitotic_count', 'oligonucleotide_primer_pair_laboratory_procedure_performed_name', 'radiation_type_notes', 'number_of_successful_pregnancies_which_resulted_in_at_least_1_live_birth', 'external_beam_radiation_therapy_administered_paraaortic_region_lymph_node_dose', 'concurrent_chemotherapy_dose', 'chemotherapy_regimen_type', 'chemotherapy_negation_radiation_therapy_concurrent_not_administered_reason', 'postoperative_rx_tx.1', 'dose_frequency_text', 'rt_pelvis_administered_total_dose', 'cervical_neoplasm_pathologic_margin_involved_type', 'cervical_neoplasm_pathologic_margin_involved_text', 'cervical_carcinoma_pelvic_extension_text', 'cervical_carcinoma_corpus_uteri_involvement_indicator', 'ectopic_pregnancy_count', 'rt_administered_type', 'oligonucleotide_primer_pair_laboratory_procedure_performed_text', 'other_chemotherapy_agent_administration_specify', 'radiation_therapy_not_administered_reason', 'patient_pregnancy_therapeutic_abortion_count', 'days_to_laboratory_procedure_tumor_marker_squamous_cell_carcinoma_antigen_result', 'days_to_performance_status_assessment', 'prescribed_dose_units', 'diagnostic_ct_result_outcome', 'patient_pregnancy_spontaneous_abortion_count', 'patient_pregnancy_count', 'fdg_or_ct_pet_performed_outcome', 'female_breast_feeding_or_pregnancy_status_indicator', 'standardized_uptake_value_cervix_uteri_assessment_measurement', 'pregnant_at_diagnosis', 'pregnancy_stillbirth_count', 'patient_history_immune_system_and_related_disorders_text', 'patient_history_immune_system_and_related_disorders_name', 'days_to_fdg_or_ct_pet_performed', 'diagnostic_mri_result_outcome', 'radiation_therapy.1', 'radiation_therapy_not_administered_specify', 'human_papillomavirus_laboratory_procedure_performed_name', 'chemotherapy_negation_radiation_therapy_concurrent_administered_text', 'lymph_node_location_positive_pathology_name', 'keratinizing_squamous_cell_carcinoma_present_indicator', 'laboratory_procedure_tumor_marker_squamous_cell_carcinoma_antigen_result_value', 'total_number_of_pregnancies', 'hysterectomy_Performed_Ind-3', 'hysterectomy_performed_text', 'human_papillomavirus_other_type_text', 'hysterectomy_performed_type', 'lymphovascular_invasion_indicator', 'brachytherapy_first_reference_point_administered_total_dose', 'brachytherapy_method_other_specify_text', 'lymph_node_location_positive_pathology_text', 'tumor_response_cdus_type', 'brachytherapy_method_type', 'agent_total_dose_count', 'live_birth_number', 'human_papillomavirus_laboratory_procedure_performed_text', 'human_papillomavirus_type', 'performance_status_assessment_timepoint_category_other_text', 'ct_scan', 'new_neoplasm_event_occurrence_anatomic_site', 'radiologic_tumor_width', 'contiguous_organ_resection_site', 'radiologic_tumor_depth', 'radiologic_tumor_burden', 'tumor_depth', 'pathologic_tumor_width', 'tumor_multifocal', 'other_contiguous_organ_resection_site', 'contiguous_organ_invaded', 'pathologic_tumor_length', 'pathologic_tumor_depth', 'pathologic_tumor_burden', 'radiologic_tumor_length', 'tumor_total_necrosis_percent', 'primary_tumor_lower_uterus_segment', 'days_to_well_differentiated_liposarcoma_resection', 'metastatic_site_at_diagnosis_other', 'metastatic_site_at_diagnosis', 'ss18_ssx_testing_method', 'ss18_ssx_fusion_status', 'specific_tumor_total_necrosis_percent', 'leiomyosarcoma_histologic_subtype', 'metastatic_neoplasm_confirmed', 'local_disease_recurrence', 'days_to_well_differentiated_liposarcoma_primary_dx', 'mpnst_neurofibromatosis', 'mpnst_exisiting_plexiform_neurofibroma', 'mpnst_nf1_genetic_testing', 'mpnst_specific_mutations', 'well_differentiated_liposarcoma_primary_dx', 'leiomyosarcoma_major_vessel_involvement', 'discontiguous_lesion_count', 'mpnst_neurofibromatosis_heredity', 'maximum_tumor_dimension', 'alcoholic_exposure_category', 'planned_surgery_status', 'esophageal_tumor_involvement_site', 'esophageal_tumor_cental_location', 'lymph_node_metastasis_radiographic_evidence', 'days_to_pancreatitis_onset', 'days_to_diabetes_onset', 'adenocarcinoma_invasion', 'surgery_performed_type', 'columnar_metaplasia_present', 'relative_cancer_type', 'amt_alcohol_consumption_per_day', 'columnar_mucosa_goblet_cell_present', 'history_of_chronic_pancreatitis', 'history_of_diabetes', 'goblet_cells_present', 'history_of_esophageal_cancer', 'histologic_grading_tier_category', 'columnar_mucosa_dysplasia', 'number_of_relatives_diagnosed', 'state_province_of_procurement', 'treatment_prior_to_surgery', 'initial_diagnosis_by', 'surgery_performed_type_other', 'country_of_birth', 'disease_detected_on_screening', 'history_pheo_or_para_include_benign', 'outside_adrenal', 'history_pheo_or_para_anatomic_site', 'family_member_relationship_type', 'post_op_ablation_embolization_tx', 'days_to_pre_orchi_serum_test', 'post_orchi_hcg', 'bilateral_diagnosis_timing_type', 'post_orchi_ldh', 'testis_tumor_microextent', 'undescended_testis_corrected', 'post_orchi_lh', 'post_orchi_lymph_node_dissection', 'post_orchi_testosterone', 'serum_markers', 'undescended_testis_corrected_age', 'undescended_testis_method_left', 'relation_testicular_cancer', 'postoperative_tx', 'testis_tumor_macroextent', 'synchronous_tumor_histology_pct', 'pre_orchi_afp', 'pre_orchi_hcg', 'pre_orchi_ldh', 'pre_orchi_lh', 'pre_orchi_testosterone', 'synchronous_tumor_histology_type', 'relative_family_cancer_hx_text', 'days_to_post_orchi_serum_test', 'undescended_testis_method_right', 'testis_tumor_macroextent_other', 'days_to_bilateral_tumor_dx', 'level_of_non_descent', 'post_orchi_afp', 'intratubular_germ_cell_neoplasm', 'igcccg_stage', 'history_of_undescended_testis', 'molecular_test_result', 'history_hypospadias', 'history_fertility', 'histological_percentage', 'first_treatment_success', 'family_history_testicular_cancer', 'family_history_other_cancer', 'masaoka_stage', 'history_myasthenia_gravis', 'section_myasthenia_gravis', 'new_neoplasm_occurrence_anatomic_site_text', 'percent_tumor_sarcomatoid', 'presence_of_sarcomatoid_features', 'tmem127', 'therapeutic_mitotane_lvl_recurrence', 'therapeutic_mitotane_lvl_macroscopic_residual', 'therapeutic_mitotane_levels_achieved', 'metastatic_neoplasm_initial_diagnosis_anatomic_site', 'mitoses_count', 'mitotane_therapy', 'mitotane_therapy_adjuvant_setting', 'mitotane_therapy_for_macroscopic_residual_disease', 'mitotic_rate', 'sinusoid_invasion', 'therapeutic_mitotane_lvl_progression', 'excess_adrenal_hormone_diagnosis_method_type', 'molecular_analysis_performed_indicator', 'excess_adrenal_hormone_history_type', 'necrosis', 'sdhd', 'sdhc', 'nuclear_grade_III_IV', 'sdhb', 'sdhaf2_sdh5', 'sdha', 'atypical_mitotic_figures', 'ct_scan_findings', 'ret', 'cytoplasm_presence_less_than_equal_25_percent', 'weiss_venous_invasion', 'weiss_score', 'vhl', 'invasion_of_tumor_capsule', 'diffuse_architecture', 'serum_mesothelin_lower_limit', 'asbestos_exposure_type', 'pleurodesis_performed_90_days', 'mesothelioma_detection_method', 'history_asbestos_exposure', 'pleurodesis_performed_prior', 'family_history_cancer_type', 'family_history_cancer_type_other', 'asbestos_exposure_age', 'serum_mesothelin_prior_tx', 'asbestos_exposure_years', 'serum_mesothelin_upper_limit', 'asbestos_exposure_source', 'suv_of_pleura_max', 'primary_occupation', 'primary_occupation_years_worked', 'primary_occupation_other', 'creatinine_norm_range_lower', 'creatinine_norm_range_upper', 'asbestos_exposure_age_last', 'creatinine_prior_tx', 'tumor_basal_diameter', 'tumor_basal_diameter_mx', 'metastatic_site', 'tumor_thickness', 'microvascular_density_mvd', 'tumor_shape_pathologic_clinical', 'tumor_infiltrating_lymphocytes', 'tumor_morphology_percentage', 'tumor_thickness_measurement', 'tumor_infiltrating_macrophages', 'pet_ct_standardized_values', 'eye_color', 'other_metastatic_site', 'extravascular_matrix_patterns', 'histological_type.1', 'extranocular_nodule_size', 'extrascleral_extension', 'gene_expression_profile', 'cytogenetic_abnormality', 'result_of_immunophenotypic_analysis', 'prior_immunosuppressive_therapy_other', 'prior_immunosuppressive_therapy_type', 'prior_infectious_disease', 'igk_genotype_results.1', 'immunophenotypic_analysis_method', 'igk_genotype_results', 'bone_marrow_biopsy_done', 'bone_marrow_involvement', 'bone_marrow_involvement.1', 'prior_infectious_disease_other', 'ebv_positive_malignant_cells_percent', 'b_lymphocyte_genotyping_method.1', 'prior_immunologic_disease_type', 'igh_genotype_results.1', 'b_lymphocyte_genotyping_method', 'prior_immunologic_disease_other', 'maximum_tumor_bulk_anatomic_location', 'immunophenotypic_analysis_results', 'maximum_tumor_bulk_anatomic_site', 'immunophenotypic_analysis_test', 'extranodal_sites_involvement_number', 'immunophenotypic_analysis_tested', 'ebv_status_malignant_cells_method', 'percent_ebv_positive_malignant_cells', 'tumor_resected_max_dimension', 'immunophenotypic_analysis_methodology', 'igh_genotype_results', 'ebv_diagnostic_methodology', 'follicular_component_percent', 'abnormality_tested_methodology', 'nodal_anatomic_site', 'histology_of_bone_marrow_sample', 'abnormality_tested_methodology.1', 'ldh_upper_limit', 'number_of_involved_extranodal_sites', 'extranodal_disease_involvement_site', 'ebv_antibody_status', 'extranodal_disease_involvement_site_other', 'genetic_abnormality_tested_other', 'genetic_abnormality_tested.1', 'genetic_abnormality_tested', 'genetic_abnormality_results_other', 'genetic_abnormality_results', 'percentage_of_follicular_component', 'pet_scan_results', 'other_genetic_abnormality_tested', 'epstein_barr_viral_status', 'lymph_node_involvement_site', 'mib1_positive_percentage_range', 'mib1_positive_percentage_range.1', 'genetic_abnormality_method_other', 'extranodal_involvment_site_other', 'ldh_lab_value', 'hiv_positive_status', 'ldh_level', 'ldh_norm_range_upper', 'abnormality_tested_results.1', 'abnormality_tested_results', 'genetic_abnormality_method', 'bone_marrow_sample_histology', 'albumin_result_upper_limit', 'vascular_tumor_cell_type', 'albumin_result_lower_limit', 'albumin_result_specified_value', 'platelet_result_upper_limit', 'intern_norm_ratio_upper_limit', 'cholangitis_tissue_evidence', 'relative_family_cancer_history', 'platelet_result_lower_limit', 'hist_hepato_carcinoma_risk', 'hist_hepato_carc_fact', 'cancer_first_degree_relative', 'creatinine_lower_level', 'prothrombin_time_result_value', 'ca_19_9_level_upper', 'ca_19_9_level_lower', 'ca_19_9_level', 'creatinine_upper_limit', 'surgical_procedure_name', 'creatinine_value_in_mg_dl', 'bilirubin_lower_limit', 'total_bilirubin_upper_limit', 'platelet_result_count', 'fibrosis_ishak_score', 'fetoprotein_outcome_value', 'fetoprotein_outcome_upper_limit', 'fetoprotein_outcome_lower_limit', 'inter_norm_ratio_lower_limit', 'family_cancer_type_txt', 'bilirubin_upper_limit', 'days_to_last_known_alive']\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/vfriedl/anaconda/lib/python3.6/site-packages/IPython/core/interactiveshell.py:2717: DtypeWarning: Columns (8,9,27,42,44,45,46,47,48,49,50,51,52,53,55,56,57,58,59,60,61,62,63,64,65,66,67,68,69,70,71,72,73,74,75,76,77,78,79,80,81,82,83,84,85,86,87,88,89,90,91,92,93,94,95,96,97,98,99,100,101,102,103,104,105,106,107,108,109,110,111,112,113,114,115,116,117,118,119,120,121,122,123,124,125,126,127,128,129,130,131,132,133,134,135,136,137,138,139,140,141,142,143,144,145,146,147,148,149,150,151,152,153,154,155,156,157,158,159,160,161,162,163,164,165,166,167,168,169,170,171,172,173,174,175,176,177,178,179,180,181,182,183,184,185,186,187,188,189,190,191,192,193,194,195,196,197,198,199,200,201,202,203,204,205,206,207,208,209,210,211,212,213,214,215,216,217,218,219,220,221,222,223,224,225,226,227,228,229,230,231,232,233,234,235,236,237,238,239,240,241,242,243,244,245,246,247,248,249,250,251,252,253,254,255,256,257,258,259,260,261,262,263,264,265,266,267,268,269,270,271,272,273,274,275,276,277,278,279,280,281,282,283,284,285,286,287,288,289,290,291,292,293,294,295,296,297,298,299,300,301,302,303,304,305,306,307,308,309,310,311,312,313,314,315,316,317,318,319,320,321,322,323,324,325,326,327,328,329,330,333,334,335,336,338,339,340,341,342,343,344,345,346,347,348,349,350,351,352,353,354,355,356,357,358,359,360,361,362,363,364,365,366,367,368,369,370,371,372,373,374,375,376,377,378,379,380,381,382,383,384,385,386,387,388,389,390,391,392,393,394,395,396,397,398,399,400,401,402,403,404,405,406,407,408,409,410,411,412,413,414,415,416,417,418,419,420,421,422,423,424,425,426,427,428,429,430,431,432,433,434,435,436,437,438,439,440,441,442,443,444,445,446,447,448,449,450,451,452,453,454,455,456,457,458,459,460,461,462,463,464,465,466,467,468,469,470,471,472,473,474,475,476,477,478,479,480,481,482,483,484,485,486,487,488,489,490,491,492,493,494,495,496,497,498,499,500,501,502,503,504,505,506,507,508,509,510,511,512,513,514,515,516,517,518,519,520,521,522,523,524,525,526,527,528,529,530,531,532,533,534,535,536,537,538,539,540,541,542,543,544,545,546,547,548,549,550,551,552,553,554,555,556,557,558,559,560,561,562,563,564,565,566,567,568,569,570,571,572,573,574,575,576,577,578,579,580,581,582,583,584,585,586,587,588,589,590,591,592,593,594,595,596,597,598,599,600,601,602,603,604,605,606,607,608,609,610,611,612,613,614,615,616,617,618,619,620,621,622,623,624,625,626,627,628,629,630,631,632,633,634,635,636,637,639,640,641,642,643,644,645,646,647,648,649,650,651,652,653,654,655,656,657,658,659,660,661,662,663,664,665,666,667,668,669,670,671,672,673,674,675,676,677,678,679,680,681,682,683,684,685,686,687,688,689,690,691,692,693,694,695,696,697,698,699,700,701,702,703,704,705,706,707,708,709,710,711,712,713,714,715,716,717,718,719,720,721,722,723,724,725,726,727,728,729,730,731,732,733,734,735,736,737,738,739,740,741,742,743,744) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  interactivity=interactivity, compiler=compiler, result=result)\n"
     ]
    }
   ],
   "source": [
    "tcga_clinical_filename = '../../../../TCGA_Pancanatlas_Clinical_out.tsv'\n",
    "tcga_clinical_data = pd.read_csv(tcga_clinical_filename, sep='\\t', header=0, index_col=0,encoding ='latin1')\n",
    "\n",
    "print(list(tcga_clinical_data))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### TCGA ID mapping"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-05-29T21:49:05.291326Z",
     "start_time": "2018-05-29T21:49:04.887929Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(12873, 8)\n"
     ]
    }
   ],
   "source": [
    "tcga_id_mapping_filename = '../../../../TCGA_id_mapping.tsv'\n",
    "tcga_id_map = pd.DataFrame.from_csv(tcga_id_mapping_filename, sep='\\t', header=0)\n",
    "print(tcga_id_map.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Samples in the different data types"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 250,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-06-05T01:21:47.746589Z",
     "start_time": "2018-06-05T01:21:37.194895Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "10510\n",
      "10713\n",
      "5498\n",
      "11061\n",
      "11117\n",
      "12384\n",
      "12747\n"
     ]
    }
   ],
   "source": [
    "# Mutation data\n",
    "samples_in_mutation_data = list(set(tcga_id_map.loc[tcga_id_map[\"SNVs_id\"].isin(set(tcga_mutation_data[\"Tumor_Sample_Barcode\"]))\n",
    "                                          ,\"sample_id\"]))\n",
    "print(len(samples_in_mutation_data))\n",
    "\n",
    "\n",
    "# CNV data\n",
    "samples_in_cnv_data = list(set(tcga_id_map.loc[tcga_id_map[\"CNV_id\"].isin(list(tcga_cnv_data)[2:])\n",
    "                                          ,\"sample_id\"]))\n",
    "print(len(samples_in_cnv_data))\n",
    "\n",
    "# Fusion data\n",
    "samples_in_fusion_data = list(set([x for x in samples_in_fusion_data if x in list(tcga_id_map[\"sample_id\"])]))\n",
    "print(len(samples_in_fusion_data))\n",
    "# use sample IDs from expression data as a background list (PRADA is working on RNA-Seq data)\n",
    "samples_in_expression_data = list(set([str(i)[:15] for i in tcga_id_map[\"mRNA_id\"]]))\n",
    "print(len(samples_in_expression_data))\n",
    "samples_in_fusion_data = list(set(samples_in_fusion_data + samples_in_expression_data))\n",
    "print(len(samples_in_fusion_data))\n",
    "\n",
    "\n",
    "# Clinical data\n",
    "samples_in_clinical_data = list(set(tcga_id_map.loc[tcga_id_map[\"Clinical_id\"].isin(set(tcga_clinical_data[\"bcr_patient_barcode\"]))\n",
    "                                          ,\"sample_id\"]))\n",
    "print(len(samples_in_clinical_data))\n",
    "\n",
    "\n",
    "# All sample IDs\n",
    "set_of_tcga_samples = set(tcga_id_map[\"sample_id\"])\n",
    "set_of_tcga_samples = [x for x in set_of_tcga_samples if x == x] #excludes 'nan' from list\n",
    "print(len(set_of_tcga_samples))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Functions defining binary biomarker labels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 214,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-06-05T00:21:13.738751Z",
     "start_time": "2018-06-05T00:21:11.901848Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Counter({'Intron': 11040579, 'Missense_Mutation': 2888649, 'Frame_Shift_Del': 1622149, 'IGR': 1271910, 'Silent': 1238725, 'RNA': 1034254, \"3'UTR\": 800627, \"3'Flank\": 752922, \"5'Flank\": 712387, 'Frame_Shift_Ins': 404988, \"5'UTR\": 246952, 'Nonsense_Mutation': 237134, 'Splice_Site': 133921, 'In_Frame_Del': 59425, 'In_Frame_Ins': 33462, 'Translation_Start_Site': 4310, 'Nonstop_Mutation': 3176, 'Targeted_Region': 57})\n"
     ]
    }
   ],
   "source": [
    "# Define what is a nonsynonymous mutation\n",
    "print(Counter(tcga_mutation_data[\"Variant_Classification\"]))\n",
    "\n",
    "non_synonymous_mutations = {'Translation_Start_Site','Missense_Mutation','Nonsense_Mutation','Nonstop_Mutation'\n",
    "                            ,'In_Frame_Ins','In_Frame_Del','Frame_Shift_Ins','Frame_Shift_Del'\n",
    "                            , 'Splice_Site'}\n",
    "\n",
    "indel_mutations = {'In_Frame_Ins','In_Frame_Del','Frame_Shift_Ins','Frame_Shift_Del'}\n",
    "insertion_mutations = {'In_Frame_Ins','Frame_Shift_Ins'}\n",
    "deletion_mutations = {'In_Frame_Del','Frame_Shift_Del'}\n",
    "\n",
    "truncating_mutations = {'Nonsense_Mutation','Frame_Shift_Ins','Frame_Shift_Del'}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 251,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-06-05T01:22:24.373292Z",
     "start_time": "2018-06-05T01:22:23.701382Z"
    },
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def get_nonsyn_mutations(gene_name):\n",
    "    \"\"\"\n",
    "    get all nonsynonymous mutations for a specific gene from the TCGA data\n",
    "    returns a dictionary with sample_name:0/1 (0=not mutated, 1=at least one non-synonymous mutation)\n",
    "    \"\"\"\n",
    "    mutation_dict = dict([ (sample_name, None) for sample_name in set_of_tcga_samples ])\n",
    "    for sample in samples_in_mutation_data:\n",
    "        mutation_dict[sample] = 0\n",
    "    \n",
    "    for k in tcga_mutation_data.loc[(tcga_mutation_data['Hugo_Symbol'] == gene_name)\n",
    "                                & (tcga_mutation_data['Variant_Classification'].isin(non_synonymous_mutations))\n",
    "                                ,\"Tumor_Sample_Barcode\"]:\n",
    "        if k[:15] in mutation_dict:\n",
    "            mutation_dict[k[:15]] = 1\n",
    "\n",
    "    return mutation_dict\n",
    "\n",
    "def get_truncating_mutations(gene_name):\n",
    "    \"\"\"\n",
    "    get all nonsynonymous mutations for a specific gene from the TCGA data\n",
    "    returns a dictionary with sample_name:0/1 (0=not mutated, 1=at least one non-synonymous mutation)\n",
    "    \"\"\"\n",
    "    mutation_dict = dict([ (sample_name, None) for sample_name in set_of_tcga_samples ])\n",
    "    for sample in samples_in_mutation_data:\n",
    "        mutation_dict[sample] = 0\n",
    "    \n",
    "    for k in tcga_mutation_data.loc[(tcga_mutation_data['Hugo_Symbol'] == gene_name)\n",
    "                                & (tcga_mutation_data['Variant_Classification'].isin(truncating_mutations))\n",
    "                                ,\"Tumor_Sample_Barcode\"]:\n",
    "        mutation_dict[k[:15]] = 1\n",
    "\n",
    "    return mutation_dict\n",
    "\n",
    "def get_nonsyn_mutations_in_exons(gene_name, exon_list):\n",
    "    \"\"\"\n",
    "    get all nonsynonymous mutations for a specific gene from the TCGA data\n",
    "    returns a dictionary with sample_name:0/1 (0=not mutated, 1=at least one non-synonymous mutation)\n",
    "    \"\"\"\n",
    "    mutation_dict = dict([ (sample_name, None) for sample_name in set_of_tcga_samples ])\n",
    "    for sample in samples_in_mutation_data:\n",
    "        mutation_dict[sample] = 0\n",
    "\n",
    "    for k in tcga_mutation_data.loc[(tcga_mutation_data['Hugo_Symbol'] == \"KIT\")\n",
    "                                & (tcga_mutation_data['Variant_Classification'].isin(non_synonymous_mutations))\n",
    "                                & ([x.startswith(tuple(exon_list)) for x in tcga_mutation_data['Exon_Number']])\n",
    "                                ,\"Tumor_Sample_Barcode\"]:\n",
    "        mutation_dict[k[:15]] = 1\n",
    "\n",
    "    return mutation_dict\n",
    "\n",
    "def get_indel_mutations_in_exons(gene_name, exon_list, keyword='indel'):\n",
    "    \"\"\"\n",
    "    get all nonsynonymous mutations for a specific gene from the TCGA data\n",
    "    returns a dictionary with sample_name:0/1 (0=not mutated, 1=at least one non-synonymous mutation)\n",
    "    \"\"\"\n",
    "    mutation_dict = dict([ (sample_name, None) for sample_name in set_of_tcga_samples ])\n",
    "    for sample in samples_in_mutation_data:\n",
    "        mutation_dict[sample] = 0\n",
    "\n",
    "    if keyword == 'indel':\n",
    "        for k in tcga_mutation_data.loc[(tcga_mutation_data['Hugo_Symbol'] == \"KIT\")\n",
    "                                & (tcga_mutation_data['Variant_Classification'].isin(indel_mutations))\n",
    "                                & ([x.startswith(tuple(exon_list)) for x in tcga_mutation_data['Exon_Number']])\n",
    "                                ,\"Tumor_Sample_Barcode\"]:\n",
    "            mutation_dict[k[:15]] = 1\n",
    "    elif keyword == 'insertion':\n",
    "        for k in tcga_mutation_data.loc[(tcga_mutation_data['Hugo_Symbol'] == \"KIT\")\n",
    "                                & (tcga_mutation_data['Variant_Classification'].isin(insertion_mutations))\n",
    "                                & ([x.startswith(tuple(exon_list)) for x in tcga_mutation_data['Exon_Number']])\n",
    "                                ,\"Tumor_Sample_Barcode\"]:\n",
    "            mutation_dict[k[:15]] = 1\n",
    "\n",
    "    elif keyword == 'deletion':\n",
    "        for k in tcga_mutation_data.loc[(tcga_mutation_data['Hugo_Symbol'] == \"KIT\")\n",
    "                                & (tcga_mutation_data['Variant_Classification'].isin(deletion_mutations))\n",
    "                                & ([x.startswith(tuple(exon_list)) for x in tcga_mutation_data['Exon_Number']])\n",
    "                                ,\"Tumor_Sample_Barcode\"]:\n",
    "            mutation_dict[k[:15]] = 1\n",
    "            \n",
    "    else:\n",
    "        print(\"no valid keyword provided: has to be one of [indel,insertion,deletion]\")\n",
    "\n",
    "    return mutation_dict\n",
    "\n",
    "def get_any_mutations(gene_name):\n",
    "    \"\"\"\n",
    "    get all mutations for a specific gene from the TCGA data\n",
    "    returns a dictionary with sample_name:0/1 (0=not mutated, 1=at least one mutation)\n",
    "    \"\"\"\n",
    "    mutation_dict = dict([ (sample_name, None) for sample_name in set_of_tcga_samples ])\n",
    "    for sample in samples_in_mutation_data:\n",
    "        mutation_dict[sample] = 0\n",
    "\n",
    "    for k in tcga_mutation_data.loc[(tcga_mutation_data['Hugo_Symbol'] == gene_name)\n",
    "                                ,\"Tumor_Sample_Barcode\"]:\n",
    "        mutation_dict[k[:15]] = 1\n",
    "        \n",
    "    return mutation_dict\n",
    "\n",
    "def get_point_mutations(gene_name,list_of_AA_changes):\n",
    "    \n",
    "    point_mut_dict = dict([ (sample_name, None) for sample_name in set_of_tcga_samples ])\n",
    "    for sample in samples_in_mutation_data:\n",
    "        point_mut_dict[sample] = 0\n",
    "    \n",
    "    list_of_AA_changes = [\"p.\"+change for change in list_of_AA_changes]\n",
    "    for k in tcga_mutation_data.loc[(tcga_mutation_data[\"Hugo_Symbol\"]==gene_name)\n",
    "                       &(tcga_mutation_data[\"HGVSp_Short\"].isin(list_of_AA_changes))\n",
    "                       ,\"Tumor_Sample_Barcode\"]:\n",
    "        point_mut_dict[k[:15]] = 1\n",
    "    \n",
    "    return point_mut_dict\n",
    "\n",
    "def get_deletions(gene_name):\n",
    "    \"\"\"\n",
    "    get all deletions (-2 in the thresholded data) for a specific gene from the TCGA data\n",
    "    returns a dictionary with sample_name:0/1 (0=not deleted, 1= deleted (-2))\n",
    "    \"\"\"\n",
    "    \n",
    "    cnv_dict = dict([ (sample_name, None) for sample_name in set_of_tcga_samples ])\n",
    "    for sample in samples_in_cnv_data:\n",
    "        cnv_dict[sample] = 0\n",
    "    \n",
    "    for k in tcga_cnv_data.columns[tcga_cnv_data.loc[gene_name,] == -2]:\n",
    "        cnv_dict[k[:15]] = 1\n",
    "    \n",
    "    return cnv_dict\n",
    "\n",
    "def get_deletions_1(gene_name):\n",
    "    \"\"\"\n",
    "    get all deletions (-1 in the thresholded data) for a specific gene from the TCGA data\n",
    "    returns a dictionary with sample_name:0/1 (0=not deleted, 1= deleted (-1))\n",
    "    \"\"\"\n",
    "    cnv_dict = dict([ (sample_name, None) for sample_name in set_of_tcga_samples ])\n",
    "    for sample in samples_in_cnv_data:\n",
    "        cnv_dict[sample] = 0\n",
    "        \n",
    "    for k in tcga_cnv_data.columns[tcga_cnv_data.loc[gene_name,] == -1]:\n",
    "        cnv_dict[k[:15]] = 1 \n",
    "    \n",
    "    return cnv_dict\n",
    "\n",
    "def get_amplifications(gene_name):\n",
    "    \"\"\"\n",
    "    get all amplifications (2 in the thresholded data) for a specific gene from the TCGA data\n",
    "    returns a dictionary with sample_name:0/1 (0=not amplified, 1= amplified (2))\n",
    "    \"\"\"\n",
    "    cnv_dict = dict([ (sample_name, None) for sample_name in set_of_tcga_samples ])\n",
    "    for sample in samples_in_cnv_data:\n",
    "        cnv_dict[sample] = 0\n",
    "        \n",
    "    for k in tcga_cnv_data.columns[tcga_cnv_data.loc[gene_name,] == 2]:\n",
    "        cnv_dict[k[:15]] = 1  \n",
    "\n",
    "    return cnv_dict\n",
    "\n",
    "def get_amplifications_1(gene_name):\n",
    "    \"\"\"\n",
    "    get all amplifications (1 in the thresholded data) for a specific gene from the TCGA data\n",
    "    returns a dictionary with sample_name:0/1 (0=not amplified, 1= amplified (1))\n",
    "    \"\"\"\n",
    "    \n",
    "    cnv_dict = dict([ (sample_name, None) for sample_name in set_of_tcga_samples ])\n",
    "    for sample in samples_in_cnv_data:\n",
    "        cnv_dict[sample] = 0\n",
    "    \n",
    "    for k in tcga_cnv_data.columns[tcga_cnv_data.loc[gene_name,] == 1]:\n",
    "        cnv_dict[k[:15]] = 1   \n",
    "\n",
    "    return cnv_dict\n",
    "\n",
    "def get_fusion_genes(gene1, gene2):\n",
    "    \n",
    "    fusion_dict = dict([ (sample_name, None) for sample_name in set_of_tcga_samples ])\n",
    "    for sample in samples_in_fusion_data:\n",
    "        fusion_dict[sample] = 0\n",
    "    \n",
    "    for k in tcga_fusion_data.loc[(tcga_fusion_data['Gene_A'] == gene1)\n",
    "                                & (tcga_fusion_data['Gene_B'] == gene2)\n",
    "                                ,\"sampleId\"]:\n",
    "        if k[:15] in fusion_dict:\n",
    "            fusion_dict[k[:15]] = 1\n",
    "        \n",
    "    for k in tcga_fusion_data.loc[(tcga_fusion_data['Gene_A'] == gene2)\n",
    "                                & (tcga_fusion_data['Gene_B'] == gene1)\n",
    "                                ,\"sampleId\"]:\n",
    "        if k[:15] in fusion_dict:\n",
    "            fusion_dict[k[:15]] = 1\n",
    "    \n",
    "    return fusion_dict\n",
    "\n",
    "def get_fusion_gene(gene):\n",
    "    \n",
    "    fusion_dict = dict([ (sample_name, None) for sample_name in set_of_tcga_samples ])\n",
    "    for sample in samples_in_fusion_data:\n",
    "        fusion_dict[sample] = 0\n",
    "    \n",
    "    for k in tcga_fusion_data.loc[(tcga_fusion_data['Gene_A'] == gene)\n",
    "                                ,\"sampleId\"]:\n",
    "        if k[:15] in fusion_dict:\n",
    "            fusion_dict[k[:15]] = 1\n",
    "    \n",
    "    for k in tcga_fusion_data.loc[(tcga_fusion_data['Gene_B'] == gene)\n",
    "                                ,\"sampleId\"]:\n",
    "        if k[:15] in fusion_dict:\n",
    "            fusion_dict[k[:15]] = 1\n",
    "    \n",
    "    return fusion_dict\n",
    "\n",
    "def get_biallelic_inactivation(gene_name):\n",
    "\n",
    "    nonsyn_mutation_and_deletion = intersect_attribute([get_deletions_1(gene_name),get_nonsyn_mutations(gene_name)])\n",
    "    biallelic_inactivation = union_attribute([get_deletions(gene_name),nonsyn_mutation_and_deletion])\n",
    "    \n",
    "    return biallelic_inactivation\n",
    "\n",
    "def union_attribute(list_of_attribute_dicts):\n",
    "    # union of keys\n",
    "    keys = list()\n",
    "    for attribute in list_of_attribute_dicts:\n",
    "        keys.extend(attribute.keys())\n",
    "        keys = list(set(keys))\n",
    "\n",
    "    # Fill in 0s for non-missing data\n",
    "    union_dict = dict([ (sample_name, None) for sample_name in keys ])\n",
    "    for k in keys:\n",
    "        for attribute in list_of_attribute_dicts:\n",
    "            if k in attribute:\n",
    "                if attribute[k] == 0:\n",
    "                    union_dict[k] = 0\n",
    "    \n",
    "    # Overwrite with 1s (build union)\n",
    "    for k in keys:\n",
    "        for attribute in list_of_attribute_dicts:\n",
    "            if k in attribute:\n",
    "                if attribute[k] == 1:\n",
    "                    union_dict[k] = 1\n",
    "               \n",
    "    return union_dict\n",
    "\n",
    "def intersect_attribute(list_of_attribute_dicts):\n",
    "    # union of keys\n",
    "    keys = list()\n",
    "    for attribute in list_of_attribute_dicts:\n",
    "        keys.extend(attribute.keys())\n",
    "        keys = list(set(keys))\n",
    "    \n",
    "    intersect_dict = dict([ (sample_name, None) for sample_name in keys ])\n",
    "    \n",
    "    # Fill in 0s for non-missing data\n",
    "    for k in keys:\n",
    "        in_all_dicts = True\n",
    "        for attribute in list_of_attribute_dicts:\n",
    "            if k in attribute:\n",
    "                if attribute[k] == None:\n",
    "                    in_all_dicts = False\n",
    "        if in_all_dicts:\n",
    "            intersect_dict[k] = 0\n",
    "            \n",
    "    # Overwrite with 1s (intersection)\n",
    "    for k in keys:\n",
    "        in_all_dicts = True\n",
    "        for attribute in list_of_attribute_dicts:\n",
    "            if k in attribute:\n",
    "                if attribute[k] == 0:\n",
    "                    in_all_dicts = False\n",
    "                if attribute[k] == None:\n",
    "                    in_all_dicts = False\n",
    "        if in_all_dicts:\n",
    "            intersect_dict[k] = 1\n",
    "                    \n",
    "    return intersect_dict\n",
    "\n",
    "def reverse_attribute(attribute):\n",
    "    reverse_dict = dict([ (sample_name, None) for sample_name in attribute.keys() ])\n",
    "    for k in attribute.keys():\n",
    "        if attribute[k]==0:\n",
    "            reverse_dict[k] = 1\n",
    "        if attribute[k]==1:\n",
    "            reverse_dict[k] = 0\n",
    "    return reverse_dict\n",
    "    \n",
    "\n",
    "def how_many_positive_samples(attribute_dict):\n",
    "    return sum(1 for v in attribute_dict.values() if v==1)\n",
    "\n",
    "def how_many_negative_samples(attribute_dict):\n",
    "    return sum(1 for v in attribute_dict.values() if v==0)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Create biomarker labels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 272,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-06-05T01:41:12.105530Z",
     "start_time": "2018-06-05T01:40:53.532948Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(264, 10)\n",
      "(264, 12757)\n",
      "12747\n"
     ]
    }
   ],
   "source": [
    "# create empty columns for TCGA samples\n",
    "print(oncokb_table.shape)\n",
    "for sample in set_of_tcga_samples:\n",
    "    oncokb_table[sample] = np.nan\n",
    "\n",
    "print(oncokb_table.shape)\n",
    "print(len(set_of_tcga_samples))\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 276,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-06-05T02:48:25.520820Z",
     "start_time": "2018-06-05T01:49:30.946630Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Isoform                         ENST00000320356\n",
      "RefSeq                              NM_004456.4\n",
      "Entrez Gene ID                             2146\n",
      "Gene                                       EZH2\n",
      "Alteration                  Oncogenic Mutations\n",
      "Cancer Type       Diffuse Large B-Cell Lymphoma\n",
      "Level                                         4\n",
      "Name: 2, dtype: object\n",
      "Isoform               ENST00000298552\n",
      "RefSeq                    NM_000368.4\n",
      "Entrez Gene ID                   7248\n",
      "Gene                             TSC1\n",
      "Alteration        Oncogenic Mutations\n",
      "Cancer Type                CNS Cancer\n",
      "Level                              2A\n",
      "Name: 10, dtype: object\n",
      "Isoform                ENST00000298552\n",
      "RefSeq                     NM_000368.4\n",
      "Entrez Gene ID                    7248\n",
      "Gene                              TSC1\n",
      "Alteration         Oncogenic Mutations\n",
      "Cancer Type       Renal Cell Carcinoma\n",
      "Level                               2A\n",
      "Name: 11, dtype: object\n",
      "Isoform               ENST00000288135\n",
      "RefSeq                    NM_000222.2\n",
      "Entrez Gene ID                   3815\n",
      "Gene                              KIT\n",
      "Alteration        Oncogenic Mutations\n",
      "Cancer Type              Thymic Tumor\n",
      "Level                              2A\n",
      "Name: 20, dtype: object\n",
      "Isoform               ENST00000288135\n",
      "RefSeq                    NM_000222.2\n",
      "Entrez Gene ID                   3815\n",
      "Gene                              KIT\n",
      "Alteration        Oncogenic Mutations\n",
      "Cancer Type                  Melanoma\n",
      "Level                              2A\n",
      "Name: 21, dtype: object\n",
      "Isoform                          ENST00000288135\n",
      "RefSeq                               NM_000222.2\n",
      "Entrez Gene ID                              3815\n",
      "Gene                                         KIT\n",
      "Alteration                   Oncogenic Mutations\n",
      "Cancer Type       Gastrointestinal Stromal Tumor\n",
      "Level                                          1\n",
      "Name: 25, dtype: object\n",
      "Isoform               ENST00000380152\n",
      "RefSeq                    NM_000059.3\n",
      "Entrez Gene ID                    675\n",
      "Gene                            BRCA2\n",
      "Alteration        Oncogenic Mutations\n",
      "Cancer Type            Ovarian Cancer\n",
      "Level                               1\n",
      "Name: 50, dtype: object\n",
      "Isoform               ENST00000380152\n",
      "RefSeq                    NM_000059.3\n",
      "Entrez Gene ID                    675\n",
      "Gene                            BRCA2\n",
      "Alteration        Oncogenic Mutations\n",
      "Cancer Type            Ovarian Cancer\n",
      "Level                              2A\n",
      "Name: 51, dtype: object\n",
      "Isoform               ENST00000377045\n",
      "RefSeq                    NM_001654.4\n",
      "Entrez Gene ID                    369\n",
      "Gene                             ARAF\n",
      "Alteration        Oncogenic Mutations\n",
      "Cancer Type             Histiocytosis\n",
      "Level                              3A\n",
      "Name: 57, dtype: object\n",
      "Isoform                      ENST00000377045\n",
      "RefSeq                           NM_001654.4\n",
      "Entrez Gene ID                           369\n",
      "Gene                                    ARAF\n",
      "Alteration               Oncogenic Mutations\n",
      "Cancer Type       Non-Small Cell Lung Cancer\n",
      "Level                                     3A\n",
      "Name: 58, dtype: object\n",
      "Isoform               ENST00000219476\n",
      "RefSeq                    NM_000548.3\n",
      "Entrez Gene ID                   7249\n",
      "Gene                             TSC2\n",
      "Alteration        Oncogenic Mutations\n",
      "Cancer Type                CNS Cancer\n",
      "Level                              2A\n",
      "Name: 59, dtype: object\n",
      "Isoform               ENST00000263967\n",
      "RefSeq                    NM_006218.2\n",
      "Entrez Gene ID                   5290\n",
      "Gene                           PIK3CA\n",
      "Alteration        Oncogenic Mutations\n",
      "Cancer Type             Breast Cancer\n",
      "Level                              3A\n",
      "Name: 68, dtype: object\n",
      "Isoform               ENST00000256078\n",
      "RefSeq                    NM_033360.2\n",
      "Entrez Gene ID                   3845\n",
      "Gene                             KRAS\n",
      "Alteration        Oncogenic Mutations\n",
      "Cancer Type         Colorectal Cancer\n",
      "Level                              R1\n",
      "Name: 76, dtype: object\n",
      "Isoform                      ENST00000256078\n",
      "RefSeq                           NM_033360.2\n",
      "Entrez Gene ID                          3845\n",
      "Gene                                    KRAS\n",
      "Alteration               Oncogenic Mutations\n",
      "Cancer Type       Non-Small Cell Lung Cancer\n",
      "Level                                      4\n",
      "Name: 77, dtype: object\n",
      "Isoform               ENST00000256078\n",
      "RefSeq                    NM_033360.2\n",
      "Entrez Gene ID                   3845\n",
      "Gene                             KRAS\n",
      "Alteration        Oncogenic Mutations\n",
      "Cancer Type                All Tumors\n",
      "Level                               4\n",
      "Name: 78, dtype: object\n",
      "Isoform               ENST00000206249\n",
      "RefSeq                 NM_001122740.1\n",
      "Entrez Gene ID                   2099\n",
      "Gene                             ESR1\n",
      "Alteration        Oncogenic Mutations\n",
      "Cancer Type             Breast Cancer\n",
      "Level                              3A\n",
      "Name: 85, dtype: object\n",
      "Isoform               ENST00000369535\n",
      "RefSeq                    NM_002524.4\n",
      "Entrez Gene ID                   4893\n",
      "Gene                             NRAS\n",
      "Alteration        Oncogenic Mutations\n",
      "Cancer Type                  Melanoma\n",
      "Level                              3A\n",
      "Name: 86, dtype: object\n",
      "Isoform               ENST00000369535\n",
      "RefSeq                    NM_002524.4\n",
      "Entrez Gene ID                   4893\n",
      "Gene                             NRAS\n",
      "Alteration        Oncogenic Mutations\n",
      "Cancer Type            Thyroid Cancer\n",
      "Level                              3A\n",
      "Name: 87, dtype: object\n",
      "Isoform               ENST00000369535\n",
      "RefSeq                    NM_002524.4\n",
      "Entrez Gene ID                   4893\n",
      "Gene                             NRAS\n",
      "Alteration        Oncogenic Mutations\n",
      "Cancer Type         Colorectal Cancer\n",
      "Level                              R1\n",
      "Name: 88, dtype: object\n",
      "Isoform               ENST00000357654\n",
      "RefSeq                    NM_007294.3\n",
      "Entrez Gene ID                    672\n",
      "Gene                            BRCA1\n",
      "Alteration        Oncogenic Mutations\n",
      "Cancer Type            Ovarian Cancer\n",
      "Level                               1\n",
      "Name: 89, dtype: object\n",
      "Isoform               ENST00000357654\n",
      "RefSeq                    NM_007294.3\n",
      "Entrez Gene ID                    672\n",
      "Gene                            BRCA1\n",
      "Alteration        Oncogenic Mutations\n",
      "Cancer Type            Ovarian Cancer\n",
      "Level                              2A\n",
      "Name: 90, dtype: object\n",
      "Isoform               ENST00000391945\n",
      "RefSeq                    NM_000400.3\n",
      "Entrez Gene ID                   2068\n",
      "Gene                            ERCC2\n",
      "Alteration        Oncogenic Mutations\n",
      "Cancer Type            Bladder Cancer\n",
      "Level                              3A\n",
      "Name: 93, dtype: object\n",
      "Isoform                      ENST00000277541\n",
      "RefSeq                           NM_017617.3\n",
      "Entrez Gene ID                          4851\n",
      "Gene                                  NOTCH1\n",
      "Alteration        Gain-of-function Mutations\n",
      "Cancer Type                       All Tumors\n",
      "Level                                      4\n",
      "Name: 108, dtype: object\n",
      "Isoform                      ENST00000269571\n",
      "RefSeq                           NM_004448.2\n",
      "Entrez Gene ID                          2064\n",
      "Gene                                   ERBB2\n",
      "Alteration               Oncogenic Mutations\n",
      "Cancer Type       Non-Small Cell Lung Cancer\n",
      "Level                                     3A\n",
      "Name: 110, dtype: object\n",
      "Isoform               ENST00000269571\n",
      "RefSeq                    NM_004448.2\n",
      "Entrez Gene ID                   2064\n",
      "Gene                            ERBB2\n",
      "Alteration        Oncogenic Mutations\n",
      "Cancer Type             Breast Cancer\n",
      "Level                              3A\n",
      "Name: 111, dtype: object\n",
      "Isoform           ENST00000288602\n",
      "RefSeq                NM_004333.4\n",
      "Entrez Gene ID                673\n",
      "Gene                         BRAF\n",
      "Alteration          p61BRAF-V600E\n",
      "Cancer Type              Melanoma\n",
      "Level                           4\n",
      "Name: 141, dtype: object\n",
      "Isoform               ENST00000288602\n",
      "RefSeq                    NM_004333.4\n",
      "Entrez Gene ID                    673\n",
      "Gene                             BRAF\n",
      "Alteration        Oncogenic Mutations\n",
      "Cancer Type                All Tumors\n",
      "Level                               4\n",
      "Name: 174, dtype: object\n",
      "Isoform                                 ENST00000311189\n",
      "RefSeq                                   NM_001130442.1\n",
      "Entrez Gene ID                                     3265\n",
      "Gene                                               HRAS\n",
      "Alteration                          Oncogenic Mutations\n",
      "Cancer Type       Head and Neck Squamous Cell Carcinoma\n",
      "Level                                                3A\n",
      "Name: 213, dtype: object\n",
      "Isoform               ENST00000358273\n",
      "RefSeq                 NM_001042492.2\n",
      "Entrez Gene ID                   4763\n",
      "Gene                              NF1\n",
      "Alteration        Oncogenic Mutations\n",
      "Cancer Type                All Tumors\n",
      "Level                               4\n",
      "Name: 214, dtype: object\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Isoform                             ENST00000241453\n",
      "RefSeq                                  NM_004119.2\n",
      "Entrez Gene ID                                 2322\n",
      "Gene                                           FLT3\n",
      "Alteration        FLT3 internal tandem duplications\n",
      "Cancer Type                  Acute Myeloid Leukemia\n",
      "Level                                            3A\n",
      "Name: 242, dtype: object\n",
      "Isoform               ENST00000371953\n",
      "RefSeq                    NM_000314.4\n",
      "Entrez Gene ID                   5728\n",
      "Gene                             PTEN\n",
      "Alteration        Oncogenic Mutations\n",
      "Cancer Type             Breast Cancer\n",
      "Level                               4\n",
      "Name: 244, dtype: object\n",
      "Isoform               ENST00000371953\n",
      "RefSeq                    NM_000314.4\n",
      "Entrez Gene ID                   5728\n",
      "Gene                             PTEN\n",
      "Alteration        Oncogenic Mutations\n",
      "Cancer Type                All Tumors\n",
      "Level                               4\n",
      "Name: 245, dtype: object\n",
      "Isoform               ENST00000261799\n",
      "RefSeq                    NM_002609.3\n",
      "Entrez Gene ID                   5159\n",
      "Gene                           PDGFRB\n",
      "Alteration        Oncogenic Mutations\n",
      "Cancer Type                All Tumors\n",
      "Level                               4\n",
      "Name: 250, dtype: object\n",
      "Isoform               ENST00000307102\n",
      "RefSeq                    NM_002755.3\n",
      "Entrez Gene ID                   5604\n",
      "Gene                           MAP2K1\n",
      "Alteration        Oncogenic Mutations\n",
      "Cancer Type                  Melanoma\n",
      "Level                              3A\n",
      "Name: 254, dtype: object\n",
      "Isoform                           ENST00000307102\n",
      "RefSeq                                NM_002755.3\n",
      "Entrez Gene ID                               5604\n",
      "Gene                                       MAP2K1\n",
      "Alteration                    Oncogenic Mutations\n",
      "Cancer Type       Low-Grade Serous Ovarian Cancer\n",
      "Level                                          3A\n",
      "Name: 255, dtype: object\n",
      "Isoform                      ENST00000307102\n",
      "RefSeq                           NM_002755.3\n",
      "Entrez Gene ID                          5604\n",
      "Gene                                  MAP2K1\n",
      "Alteration               Oncogenic Mutations\n",
      "Cancer Type       Non-Small Cell Lung Cancer\n",
      "Level                                     3A\n",
      "Name: 256, dtype: object\n",
      "Isoform                ENST00000307102\n",
      "RefSeq                     NM_002755.3\n",
      "Entrez Gene ID                    5604\n",
      "Gene                            MAP2K1\n",
      "Alteration         Oncogenic Mutations\n",
      "Cancer Type       Histiocytic Disorder\n",
      "Level                               3A\n",
      "Name: 257, dtype: object\n",
      "Isoform                          ENST00000257290\n",
      "RefSeq                               NM_006206.4\n",
      "Entrez Gene ID                              5156\n",
      "Gene                                      PDGFRA\n",
      "Alteration                   Oncogenic Mutations\n",
      "Cancer Type       Gastrointestinal Stromal Tumor\n",
      "Level                                         2A\n",
      "Name: 262, dtype: object\n"
     ]
    }
   ],
   "source": [
    "# fill the oncokb table with biomarker labels\n",
    "for index, row in oncokb_table.iterrows():\n",
    "    \n",
    "    feature_vec = 0\n",
    "    \n",
    "    # Fusion biomarker\n",
    "    if \"Fusion\" in row[\"Alteration\"]:  \n",
    "        if row[\"Alteration\"]==\"Fusions\":\n",
    "            feature_vec = get_fusion_gene(row[\"Gene\"])\n",
    "            #print(row[:7])\n",
    "            #print(how_many_positive_samples(feature_vec))\n",
    "        else:\n",
    "            fusion_genes = row[\"Alteration\"].split(' ')[0].split('-')\n",
    "            feature_vec = get_fusion_genes(fusion_genes[0],fusion_genes[1])\n",
    "            #print(row[:7])\n",
    "            #print(how_many_positive_samples(feature_vec))\n",
    "            \n",
    "            \n",
    "    # Copy number biomarker - TODO: No deletions?\n",
    "    elif \"Amplification\" in row[\"Alteration\"]:\n",
    "        feature_vec = get_amplifications(row[\"Gene\"])\n",
    "        #print(row[:7])\n",
    "        #print(how_many_positive_samples(feature_vec))\n",
    "    elif \"Deletion\" in row[\"Alteration\"]:\n",
    "        print(row[:7])\n",
    "        \n",
    "        \n",
    "    # Point mutations\n",
    "    elif re.fullmatch(r'[A-Z][0-9]{1,5}[A-Z]', row[\"Alteration\"], flags=0):\n",
    "        feature_vec = get_point_mutations(row[\"Gene\"],[row[\"Alteration\"]])\n",
    "        #print(row[:7])\n",
    "        #print(how_many_positive_samples(feature_vec))\n",
    "\n",
    "        \n",
    "    # Exon specific mutations\n",
    "    elif \"Exon\" in row[\"Alteration\"]:\n",
    "        exon = row[\"Alteration\"].split(' ')[1]\n",
    "        if \"mutations\" in row[\"Alteration\"]:\n",
    "            feature_vec = get_nonsyn_mutations_in_exons(row[\"Gene\"],[exon])\n",
    "            #print(row[:7])\n",
    "            #print(how_many_positive_samples(feature_vec))\n",
    "                \n",
    "        elif \"insertion\" in row[\"Alteration\"] and \"deletion\" in row[\"Alteration\"]:\n",
    "            feature_vec = get_indel_mutations_in_exons(row[\"Gene\"],[exon],'indel')\n",
    "            #print(row[:7])\n",
    "            #print(how_many_positive_samples(feature_vec))\n",
    "                \n",
    "        elif \"insertion\" in row[\"Alteration\"] and not \"deletion\" in row[\"Alteration\"]:\n",
    "            feature_vec = get_indel_mutations_in_exons(row[\"Gene\"],[exon],'insertion')\n",
    "            #print(row[:7])\n",
    "            #print(how_many_positive_samples(feature_vec))\n",
    "                \n",
    "        elif \"deletion\" in row[\"Alteration\"] and not \"insertion\" in row[\"Alteration\"]:\n",
    "            feature_vec = get_indel_mutations_in_exons(row[\"Gene\"],[exon],'deletion')\n",
    "            #print(row[:7])\n",
    "            #print(how_many_positive_samples(feature_vec))\n",
    "\n",
    "    # Nonsense and Frame-shift mutations\n",
    "    elif \"Truncating\" in row[\"Alteration\"]:\n",
    "        feature_vec = get_truncating_mutations(row[\"Gene\"])\n",
    "        #print(row[:7])\n",
    "        #print(how_many_positive_samples(feature_vec))\n",
    "\n",
    "        \n",
    "    elif \"splice\" in row[\"Alteration\"]:\n",
    "        feature_vec = get_point_mutations(row[\"Gene\"],[row[\"Alteration\"]])\n",
    "        #print(row[:7])\n",
    "        #print(how_many_positive_samples(feature_vec))\n",
    "\n",
    "        \n",
    "    elif \"Wildtype\" in row[\"Alteration\"]:\n",
    "        feature_vec = reverse_attribute(get_nonsyn_mutations(row[\"Gene\"]))\n",
    "        #print(row[:7])\n",
    "        #print(how_many_positive_samples(feature_vec))\n",
    "\n",
    "        \n",
    "    # special EGFR mutations        \n",
    "    elif \"A763_Y764insFQEA\" in row[\"Alteration\"]:\n",
    "        feature_vec = get_point_mutations(row[\"Gene\"],[row[\"Alteration\"]])\n",
    "        #print(row[:7])\n",
    "        #print(how_many_positive_samples(feature_vec))\n",
    "    elif \"E709_T710delinsD\" in row[\"Alteration\"]:\n",
    "        feature_vec = get_point_mutations(row[\"Gene\"],[row[\"Alteration\"]])\n",
    "        #print(row[:7])\n",
    "        #print(how_many_positive_samples(feature_vec))\n",
    " \n",
    "\n",
    "    # TODO: check papers for mutations\n",
    "    elif \"Oncogenic Mutations\" in row[\"Alteration\"]:\n",
    "        print(row[:7])\n",
    "        \n",
    "        \n",
    "    # only very few annotations - TODO: define on mutation data!\n",
    "    elif \"Microsatellite Instability-High\" in row[\"Alteration\"]:\n",
    "        instability_dict = dict([ (sample_name, None) for sample_name in set_of_tcga_samples ])\n",
    "    \n",
    "        for k in tcga_clinical_data.loc[(tcga_clinical_data['microsatellite_instability'] == \"YES\")\n",
    "                                ,\"bcr_patient_barcode\"]:\n",
    "            matching_sample_ids = [str(x) for x in set_of_tcga_samples if k in str(x)]\n",
    "            for sid in matching_sample_ids:\n",
    "                instability_dict[sid] = 1\n",
    "                    \n",
    "        for k in tcga_clinical_data.loc[(tcga_clinical_data['microsatellite_instability'] == \"NO\")\n",
    "                                ,\"bcr_patient_barcode\"]:\n",
    "            matching_sample_ids = [str(x) for x in set_of_tcga_samples if k in str(x)]\n",
    "            for sid in matching_sample_ids:\n",
    "                instability_dict[sid] = 0\n",
    "            \n",
    "        feature_vec = instability_dict\n",
    "        #print(row[:7])\n",
    "        #print(how_many_positive_samples(feature_vec))\n",
    "        \n",
    "        \n",
    "    else:\n",
    "        print(row[:7])\n",
    "            \n",
    "            \n",
    "            \n",
    "        \n",
    "    if(feature_vec):\n",
    "        for sample in set_of_tcga_samples:\n",
    "            oncokb_table.iloc[index,list(oncokb_table).index(sample)] = feature_vec[sample]\n",
    "            #if feature_vec[sample] == 1:\n",
    "            #    print(sample)\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 284,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-06-05T02:58:02.970162Z",
     "start_time": "2018-06-05T02:57:59.069118Z"
    },
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# write oncokb table to file\n",
    "oncokb_table.to_csv(\"./allActionableVariants_TCGAsamples.tsv\", sep='\\t', index=False, na_rep = \"NaN\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.1"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {
    "height": "calc(100% - 180px)",
    "left": "10px",
    "top": "150px",
    "width": "374px"
   },
   "toc_section_display": true,
   "toc_window_display": true
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
